packages/navigation/src/UIGridList/UIMasonryList.tsx

```diff
@@ -0,0 +1,498 @@
+// (savelichalex): following rule is just broken
+/* eslint-disable react/no-unused-prop-types */
+import * as React from 'react';
+import {
+    LayoutChangeEvent,
+    NativeScrollEvent,
+    View,
+    ViewStyle,
+    ScrollView,
+    NativeSyntheticEvent,
+    ScrollViewProps,
+} from 'react-native';
+
+import { UIConstant } from '../constants';
+import { wrapScrollableComponent } from '../Scrollable/wrapScrollableComponent';
+
+const interColumnOffset = UIConstant.contentOffset / 2;
+const interRowsOffset = interColumnOffset;
+
+type MasonryItem<Item> = { key: string; item?: Item; aspectRatio: number };
+
+type LayoutCell = {
+    height: number;
+    x: number;
+    y: number;
+};
+
+type Layout = Record<string, LayoutCell>;
+
+function measureList<Item>(cellWidth: number, data: MasonryItem<Item>[], numOfColumns: number) {
+    const columnsHeights = new Array(numOfColumns).fill(0);
+
+    let maxHeight = 0;
+
+    const layout: Layout = {};
+
+    // Calculating offsets for columns in advance,
+    // since for each row they are simply going to
+    // be repeated agan
+    const offsetsByColumn = new Array(numOfColumns).fill(null).map((_, column) => {
+        return (
+            UIConstant.contentOffset +
+            column * cellWidth +
+            (column > 0 ? column * interRowsOffset : 0)
+        );
+    });
+
+    for (let i = 0; i < data.length; i += 1) {
+        const item = data[i];
+        const row = Math.trunc(i / numOfColumns);
+        const column = i - row * numOfColumns;
+        const height = cellWidth * item.aspectRatio;
+
+        if (height > maxHeight) {
+            maxHeight = height;
+        }
+
+        const y = columnsHeights[column];
+        const x = offsetsByColumn[column];
+
+        columnsHeights[column] += height + interColumnOffset;
+
+        // eslint-disable-next-line no-param-reassign
+        layout[item.key] = {
+            height,
+            x,
+            y,
+        };
+    }
```

Comment:
Can you add a comment that this layout algo might change in the future as it has some cons as having one column way longer then another (i.e. when having several "long" images in one column)?